src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs

```
@@ -214,23 +255,63 @@ internal static bool ValidateSslServerCertificate(X509Certificate clientCert, X5
+
+                    if (!s_cert.MatchesHostname(c_cert.SubjectName.Name))
+                    {
+                        SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "certificate from server does not match with the certificate provided client.", args0: s_cert.Subject);
+                        throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNameMismatch);
+                    }
+#else
                     // Verify that subject name matches
                     if (serverCert.Subject != clientCert.Subject)
                     {
                         SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "certificate subject from server is {0}, and does not match with the certificate provided client.", args0: serverCert.Subject);
-                        return false;
+                        throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNameMismatch);
                     }
                     if (!serverCert.Equals(clientCert))
                     {
                         SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "certificate from server does not match with the certificate provided client.", args0: serverCert.Subject);
-                        return false;
+                        throw ADP.SSLCertificateAuthenticationException(Strings.SQL_RemoteCertificateNameMismatch);
                     }
+#endif
                 }
+                // unknown policy errors
                 else
                 {
-                    // Fail all other SslPolicy cases besides RemoteCertificateNameMismatch
                     SqlClientEventSource.Log.TrySNITraceEvent(nameof(SNICommon), EventType.ERR, "certificate subject: {0}, SslPolicyError {1}, SSL Policy invalidated certificate.", args0: clientCert.Subject, args1: policyErrors);
                     return false;
```

We should throw the same exception here as we do on line 233.